### PR TITLE
Refactor download route to support multiple file formats

### LINKED ISSUE #268: Enable download in both formats: CSV and JSON
For historic reasons we allow downloading evaluation results in CSV and datasets in JSON. There is no reason to disallow the other format, so we should allow _both_ CSV and JSON in _both_ places.

UI: I would like to make the download button a dropdown, so that when you click "Donwload" there are buttons that appear under it: "Download CSV" and "Download JSON", somewhat like Supabase's "Export button".

Screenshot 2024-12-05 at 17 32 11

Pointers:
- Bare JSON download in dataset: https://github.com/lmnr-ai/lmnr/blob/main/frontend/app/api/projects/%5BprojectId%5D/datasets/%5BdatasetId%5D/download/route.ts
- CSV conversion before download in https://github.com/lmnr-ai/lmnr/blob/main/frontend/app/api/projects/%5BprojectId%5D/evaluations/%5BevaluationId%5D/download/route.ts#L69-L72

### PR DESCRIPTION
**Pull request by devansh-m12**
fixes #268 

Added for evaluation. @dinmukhamedm, could you share a sample CSV for the dataset? The sample dataset I'm currently using is: [https://gist.github.com/dinmukhamedm/f90286065c3201e6c75fc7d2e3fda704](https://gist.github.com/dinmukhamedm/f90286065c3201e6c75fc7d2e3fda704).

https://github.com/user-attachments/assets/5205aaac-5cb5-4251-908a-10d410a79917

----

> [!IMPORTANT]
> Refactor download routes and components to support multiple file formats (JSON, CSV) with format validation and selection.
> 
>   - **Routes**:
>     - Refactor `download/route.ts` to `download/[format]/route.ts` for datasets and evaluations.
>     - Add `DownloadFormat` enum to validate supported formats (JSON for datasets, JSON and CSV for evaluations).
>     - Return 400 error for unsupported formats.
>   - **Components**:
>     - Update `DownloadButton` in `download-button.tsx` to support format selection via dropdown.
>     - Modify `dataset.tsx` and `evaluation.tsx` to use updated `DownloadButton` with `supportedFormats` prop.
>   - **Misc**:
>     - Remove `fileFormat` prop from `DownloadButton` and replace with `supportedFormats` array.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=lmnr-ai%2Flmnr&utm_source=github&utm_medium=referral)<sup> for e70804f3688eb271d0e4fefd8000c4c95b8c9b49. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: frontend/app/api/projects/[projectId]/evaluations/[evaluationId]/download/[format]/route.ts [hunk 1]
```diff
@@ -9,14 +9,14 @@ export async function GET(
   {
     params
   }: {
-    params: { projectId: string; evaluationId: string; };
+    params: { projectId: string; evaluationId: string; format: string };
```
Comment (dinmukhamedm) [inline frontend/app/api/projects/[projectId]/evaluations/[evaluationId]/download/[format]/route.ts]: let's make format a Enum

File: frontend/components/ui/download-button.tsx [hunk 2]
```diff
@@ -44,32 +52,42 @@ const downloadFile = async (
...
-        await downloadFile(uri, fileFormat, filenameFallback);
-        setIsDownloading(false);
-      }}
-    >
-      {isDownloading && <Loader2 className="animate-spin h-4 w-4 mr-2" />}
-      Download {fileFormat}
-    </Button>
+    <DropdownMenu>
+      <DropdownMenuTrigger className="flex h-7 w-full items-center justify-between rounded-md border bg-transparent px-3 py-2 text-sm focus:outline-none">
+        <span>Download</span>
```
Comment (dinmukhamedm) [inline frontend/components/ui/download-button.tsx]: I don't think you need <span> here

File: frontend/components/ui/download-button.tsx [hunk 3]
```diff
@@ -44,32 +52,42 @@ const downloadFile = async (
...
+        <CaretSortIcon className="h-4 w-4 opacity-50" />
```
Comment (dinmukhamedm) [inline frontend/components/ui/download-button.tsx]: Can we make it just the arrow down, not up down? Also, why the opacity? Not saying opacity is wrong, just curious

File: frontend/components/ui/download-button.tsx (hunk 2 above)
Comment (devansh-m12) [reply frontend/components/ui/download-button.tsx:71]: wdym? could you explain further

Comment (dinmukhamedm) [reply frontend/components/ui/download-button.tsx]: Oops, I think github doesn't render the html tag span when you don't wrap it in backticks. Something to do with ghm for sure haha.

I don't think you need `<span>` there

File: frontend/components/ui/download-button.tsx (hunk 3 above)
Comment (devansh-m12) [reply frontend/components/ui/download-button.tsx:72]: here, opacity is to sync the icon ui with rest of them.

Screenshot 2024-12-06 at 11 00 38 PM

**Comment by devansh-m12** [comment]
@dinmukhamedm added requested changes and here the ss for dataset download
Screenshot 2024-12-06 at 11 21 20 PM

File: frontend/app/api/projects/[projectId]/datasets/[datasetId]/download/[format]/route.ts [hunk 4]
```diff
@@ -3,18 +3,31 @@ import { and, asc, eq } from 'drizzle-orm';
 import { db } from '@/lib/db/drizzle';
 import { datasetDatapoints, datasets } from '@/lib/db/migrations/schema';
 
+enum DownloadFormat {
+  JSON = 'json'
+}
+
```
Comment (dinmukhamedm) [inline frontend/app/api/projects/[projectId]/datasets/[datasetId]/download/[format]/route.ts]: let's use a single enum, somewhere in `lib/types` for example. Also, why does this route not support CSV? Let's allow downloading as CSV but just make it have three columns with stringified JSONs each – sorry maybe I was not clear enough in the prev comment

File: frontend/components/ui/download-button.tsx [hunk 5]
```diff
@@ -1,5 +1,5 @@
-import { CaretSortIcon } from "@radix-ui/react-icons";
 import { Loader2 } from "lucide-react";
+import { ArrowDownIcon } from "@radix-ui/react-icons";
```
Comment (dinmukhamedm) [inline frontend/components/ui/download-button.tsx]: Let's use 'lucide-react' only. Let's use `ChevronDown`. https://lucide.dev/icons/chevron-down

**Review by dinmukhamedm — CHANGES_REQUESTED** [review]

**Comment by devansh-m12** [comment]
here @dinmukhamedm  the sample csv for dataset after download

[test-92686b44-9ba6-426d-9521-b3cda93a74f7 (1).csv](https://github.com/user-attachments/files/18042216/test-92686b44-9ba6-426d-9521-b3cda93a74f7.1.csv)

File: frontend/app/api/projects/[projectId]/evaluations/[evaluationId]/download/[format]/route.ts [hunk 6]
```diff
@@ -78,7 +74,7 @@ export async function GET(
     };
   });
 
-  // if the format is json
+  // else the format is json, return the results as json
```
Comment (dinmukhamedm) [inline frontend/app/api/projects/[projectId]/evaluations/[evaluationId]/download/[format]/route.ts:77]: nit: this should probably say `if`, not `else` :)

**Review by dinmukhamedm — COMMENTED** [review]
just in case, can you make sure you run `pnpm lint:fix` and `pnpm build`. If that changes any files, commit the changes too. LGTM otherwise

**Review by dinmukhamedm — APPROVED** [review]

**Pull request merged**